File: src/demo/FireWorks/FireWorks.rb
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;


    public partial class FireWorks : Form
    {
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        public static void Main()
        {
            Application.Run(new FireWorks());
        }

        public FireWorks()
        {
            InitializeComponent();
        }

        public void Add(PictureBox p)
        {
            this.Controls.Add(p);
        }

        public static int        SPARK_WIDTH      = 19;
        public static int        SPARK_HEIGHT     = 20;
        public static int        SPARK_COUNT      = 20;
        public static int        BOMB_COUNT       = 4;

        public static int        WINDOW_WIDTH     = 800;
        public static int        WINDOW_HEIGHT    = 600;

        public const float              G                = 9.81f;       // m/(s*s)
        public const float              t                = 0.25f;       // s
        public const float              dV               = t * G;       // m/s

        int                      counter          = 0;
        int                      pointer          = 0;

        Bomb[]                   bombs;

        private void FireWorks_Load(object sender, EventArgs e)
        {

            this.Text = "XML11 - FireWorks";
            this.Width = WINDOW_WIDTH;
            this.Height = WINDOW_HEIGHT;
            this.BackColor = Color.Black;


            
            bombs = new Bomb[BOMB_COUNT];
            for (int i = 0; i < bombs.Length; ++i)
            {
                bombs[i] = new Bomb(this);
            }


            Timer timer = new Timer();
            timer.Tick += new EventHandler(actionPerformed);
            timer.Interval = 60;
            timer.Enabled = true;
        }

        void actionPerformed(object sender, EventArgs e)
        {
            Random r = new Random();

            if (counter-- <= 0) {
                counter = 15;

      

                pointer = (pointer + 1) % bombs.Length;

                int x = ((int) (r.NextDouble() * 600.0)) + 100;
                int y = ((int) (r.NextDouble() * 400.0));
                bombs[pointer].init(x, y);
            }

            for (int i = 0; i < bombs.Length; ++i) {
                bombs[i].calcNextStep(t, dV);
            }
        }

    }


    class Spark
    {

        public PictureBox panel;
        public float x;
        public float y;
        public float vx;
        public float vy;
    }
    class Bomb
    {

        Spark[] sparks;
        FireWorks parent;

        public Bomb()
        {
        }
        // TODO: Auto gen
        public static Bomb __NEW(FireWorks parent)
        {
            Bomb toRet = new Bomb();
            InitBomb(toRet,parent);
            return toRet;
        }
        public Bomb(FireWorks parent)
        {
            InitBomb(this, parent);
        }

        static void InitBomb(Bomb b,FireWorks parent)
        {
            b.parent = parent;
            b.sparks = new Spark[FireWorks.SPARK_COUNT];


            for (int i = 0; i < b.sparks.Length; ++i)
            {
                b.sparks[i] = new Spark();
                b.sparks[i].panel = new PictureBox();
                b.sparks[i].panel.Image = Image.FromFile("spark.gif");

                //sparks[i].panel.Image = new Bitmap(typeof(FireWorks),"spark.gif");

                b.parent.Add(b.sparks[i].panel);

            }
        }


        public void init(int x, int y)
        {
            Random r = new Random();
            for (int i = 0; i < sparks.Length; ++i)
            {

                sparks[i].x = x;
                sparks[i].y = y;

                sparks[i].vx = (float)(r.NextDouble()* 60) - 30;
                sparks[i].vy = (float)(r.NextDouble()* 60) - 30;
            }
        }



        public void calcNextStep(float secondsPassed, float dV)
        {

            for (int i = 0; i < sparks.Length; ++i)
            {
                if ((sparks[i].x < FireWorks.WINDOW_WIDTH) && 
                    (sparks[i].x > -20)
                        && (sparks[i].y < FireWorks.WINDOW_WIDTH)
                        && (sparks[i].y > -20))
                {
                    
                    // Calculate new positions
                    sparks[i].vy += dV;
                    sparks[i].y += secondsPassed * sparks[i].vy;
                    sparks[i].x += secondsPassed * sparks[i].vx;

                    // Update the position
                    sparks[i].panel.SetBounds
                        ((int)sparks[i].x, (int)sparks[i].y,
                            FireWorks.SPARK_WIDTH, FireWorks.SPARK_HEIGHT);
                     
                }
            }

        }
    }

File: demo/java-desktop/org/xmlvm/demo/Fireworks.java
/* Copyright (c) 2002-2011 by XMLVM.org
 *
 * Project Info:  http://www.xmlvm.org
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 */

package org.xmlvm.demo;

import java.awt.Color;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.Frame;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import org.xmlvm.client.Timer;

/**
 * Implements a small fireworks application with randomly exploding bombs that
 * emit sparks.
 * 
 * @author Sascha Haeberling
 * 
 */
public class Fireworks implements ActionListener {
    public static final long serialVersionUID = 1;
    public static String     SPARK1_IMG       = "doc/fireworks/star1.png";
    public static String     SPARK2_IMG       = "doc/fireworks/star2.png";
    public static String     SPARK3_IMG       = "doc/fireworks/star3.png";
    public static String     SPARK4_IMG       = "doc/fireworks/star4.png";
    public static int        SPARK_WIDTH      = 19;
    public static int        SPARK_HEIGHT     = 20;
    public static int        SPARK_COUNT      = 20;
    public static int        BOMB_COUNT       = 4;

    public static int        WINDOW_WIDTH     = 800;
    public static int        WINDOW_HEIGHT    = 600;

    final float              G                = 9.81f;                    // m/(s*s)
    final float              t                = 0.55f;                    // s
    final float              dV               = t * G;                    // m/s

    int                      counter          = 0;
    int                      pointer          = 0;

    Bomb[]                   bombs;

    private Frame            frame;

    public static void main(String args[]) {
        new Fireworks();
    }

    public Fireworks() {
        frame = new Frame("XML11 - Fireworks");
        frame.setTitle("XML11 - Fireworks");
        frame.setLayout(null);
        frame.setSize(new Dimension(WINDOW_WIDTH, WINDOW_HEIGHT));
        frame.setBackground(Color.BLACK);
        frame.setVisible(true);
        frame.addWindowListener(new WindowAdapter() {
            @Override
            public void windowClosing(WindowEvent e) {
                System.exit(0);
            }
        });

        bombs = new Bomb[BOMB_COUNT];
        for (int i = 0; i < bombs.length; ++i) {
            bombs[i] = new Bomb(frame);
        }

        Timer timer = new Timer();
        timer.addActionListener(this);
        timer.runRepeating(60);
    }

    public void setLocation(int x, int y) {
        frame.setLocation(x, y);
    }

    public void actionPerformed(ActionEvent e) {
        if (counter-- <= 0) {
            counter = 15;
            pointer = (pointer + 1) % bombs.length;
            int x = ((int) (Math.random() * 600.0)) + 100;
            int y = ((int) (Math.random() * 400.0));
            bombs[pointer].init(x, y);
        }
        for (int i = 0; i < bombs.length; ++i) {
            bombs[i].calcNextStep(t, dV);
        }
    }
}

/**
 * In principle an image with a given position and speed vector.
 * 
 * @author Sascha Haeberling
 * 
 */
class Spark {
    ImagePanel   panel;
    public float x;
    public float y;
    public float vx;
    public float vy;
}

/**
 * A bomb contains a number of sparks. It also contains the logic to calculate
 * the next position of a spark.
 * 
 * @author Sascha Haeberling
 * 
 */
class Bomb {
    Spark[]   sparks;
    Container parent;

    public Bomb(Container parent_) {
        this.parent = parent_;
        sparks = new Spark[Fireworks.SPARK_COUNT];
        // try {
        for (int i = 0; i < sparks.length; ++i) {
            sparks[i] = new Spark();
            double rnd = Math.random() * 4.0;
            if (rnd < 1.0) {
                sparks[i].panel = new ImagePanel(Fireworks.SPARK1_IMG);
            } else if (rnd < 2.0) {
                sparks[i].panel = new ImagePanel(Fireworks.SPARK2_IMG);
            } else if (rnd < 3.0) {
                sparks[i].panel = new ImagePanel(Fireworks.SPARK3_IMG);
            } else {
                sparks[i].panel = new ImagePanel(Fireworks.SPARK4_IMG);
            }
            // sparks[i].panel.loadImage(Fireworks.SPARK_IMG);
            parent.add(sparks[i].panel);
        }
    }

    /**
     * Initialized a bomb at a given position
     * 
     * @param x
     * @param y
     */
    public void init(int x, int y) {
        for (int i = 0; i < sparks.length; ++i) {
            sparks[i].x = x;
            sparks[i].y = y;
            sparks[i].vx = (float) (Math.random() * 60) - 30;
            sparks[i].vy = (float) (Math.random() * 60) - 30;
        }
    }

    /**
     * Calculates the next step.
     * 
     * @param secondsPassed
     * @param dV
     */
    public void calcNextStep(float secondsPassed, float dV) {
        for (int i = 0; i < sparks.length; ++i) {
            if ((sparks[i].x < Fireworks.WINDOW_WIDTH) && (sparks[i].x > -20)
                    && (sparks[i].y < Fireworks.WINDOW_WIDTH) && (sparks[i].y > -20)) {
                // Calculate new positions
                sparks[i].vy += dV;
                sparks[i].y += secondsPassed * sparks[i].vy;
                sparks[i].x += secondsPassed * sparks[i].vx;

                // Update the position
                sparks[i].panel.setBounds((int) sparks[i].x, (int) sparks[i].y,
                        Fireworks.SPARK_WIDTH, Fireworks.SPARK_HEIGHT);
            }
        }
    }
}